app/Jobs/CloneImagesOrVideos.php

```diff
@@ -0,0 +1,480 @@
+     *
+     * @var string
+     **/
+    public string $cloneName;
+
+    /**
+     * Array containing file ids.
+     *
+     * @var array
+     **/
+    public array $onlyFiles;
+
+    /**
+     * Boolean for cloning annotation labels.
+     *
+     * @var bool
+     **/
+    public bool $cloneAnnotations;
+
+    /**
+     * Array containing annotation label ids.
+     *
+     * @var array
+     **/
+    public array $onlyAnnotationLabels;
+
+    /**
+     * Boolean for cloning file labels.
+     *
+     * @var bool
+     **/
+    public bool $cloneFileLabels;
+
+    /**
+     * Array containing file label ids.
+     *
+     * @var array
+     **/
+    public array $onlyFileLabels;
+
+
+    /**
+     * Create a new job instance.
+     *
+     * @param CloneVolume $request containing the project, volume, new volume name and ids of files and labels.
+     *
+     * @return void
+     */
+    public function __construct($request)
+    {
+        $this->project = $request->project;
+        $this->volume = $request->volume;
+        $this->cloneName = $request->input('name', $this->volume->name);
+        $this->onlyFiles = $request->input('only_files', []);
+        $this->cloneAnnotations = $request->input('clone_annotations', false);
+        $this->onlyAnnotationLabels = $request->input('only_annotation_labels', []);
+        $this->cloneFileLabels = $request->input('clone_file_labels', false);
+        $this->onlyFileLabels = $request->input('only_file_labels', []);
+
+    }
+
+    public function handle()
+    {
+        DB::transaction(function () {
+            $project = $this->project;
+            $volume = $this->volume;
+            $onlyFiles = $this->onlyFiles;
+            $cloneAnnotations = $this->cloneAnnotations;
+            $onlyAnnotationLabels = $this->onlyAnnotationLabels;
+            $cloneFileLabels = $this->cloneFileLabels;
+            $onlyFileLabels = $this->onlyFileLabels;
+
+            $copy = $volume->replicate();
+            $copy->name = $this->cloneName;
+            $copy->save();
+
+            if ($volume->isImageVolume()) {
+                $this->copyImages($volume, $copy, $onlyFiles);
+                if ($cloneAnnotations) {
+                    $this->copyImageAnnotation($volume, $copy, $onlyFiles, $onlyAnnotationLabels);
+                }
+                if ($cloneFileLabels) {
+                    $this->copyImageLabels($volume, $copy, $onlyFiles, $onlyFileLabels);
+                }
+            } else {
+                $this->copyVideos($volume, $copy, $onlyFiles);
+                if ($cloneAnnotations) {
+                    $this->copyVideoAnnotation($volume, $copy, $onlyFiles, $onlyAnnotationLabels);
+                }
+                if ($cloneFileLabels) {
+                    $this->copyVideoLabels($volume, $copy, $onlyFiles, $onlyFileLabels);
+                }
+            }
+            if ($copy->files()->exists()) {
+                ProcessNewVolumeFiles::dispatch($copy);
```

Comment:
I just remembered another thing that is missing here.

While the annotations are cloned correctly (if there are any), the biigle/largo module also needs to generate thumbnail images for the annotations. This can be done in a similar way than the postprocessing of a [volume import through the admin panel](https://github.com/biigle/sync/blob/7430b7eb898768a58842a44b173d139594d10809/src/Jobs/PostprocessVolumeImport.php#L50-L68). It checks if the job classes of the biigle/largo module are there (i.e. the module is installed) and then pushes one job for each newly imported annotation.

Similarly, [the tests](https://github.com/biigle/sync/blob/7430b7eb898768a58842a44b173d139594d10809/tests/Jobs/PostprocessVolumeImportTest.php#L26-L52) require the biigle/largo module to be installed. They use `markTestSkipped()` if biigle/largo is not there.